### ISSUE TITLE
Add "only-unassigned' option to "one-var" rule

### ISSUE THREAD

**Issue by Flet** [eslint/eslint#2206]
If someone has a better name for the option, thats OK :)

This is analogous to [jscs](http://jscs.info/rules.html)'s { "disallowMultipleVarDecl": "exceptUndefined" } option.

__Example Documentation:__

When configured with `"only-unassigned"` as the first option, the following patterns are considered warnings:

```js
function foo() {
    var bar = "hello", 
        baz = 1;
}

function foo() {
    var bar = 1, 
        qux;
}
```

The following patterns are not considered warnings:
```js
function foo() {
    var bar, baz;
    var foo = 1;
}

function foo() {
    var bar;
    var baz;
}
```

This is the continuation of a tangent from #1619

**Comment by nzakas** [comment]
Hmm, I'm not very clear on what you're proposing and "only-unassigned" doesn't really help explain it. 

Looking at the examples, it seems like maybe you're saying this option should require multiple `var` statement EXCEPT if all identifiers in the statement are uninitialized? Is that right?

**Comment by feross** [comment]
> maybe you're saying this option should require multiple var statement EXCEPT if all identifiers in the statement are uninitialized? Is that right?

Yep, that is correct.

**Comment by feross** [comment]
I can put up $20 for this issue (via Bountysource or Bitcoin, whichever you prefer). Bountysource isn't listing this issue for some reason, or I'd have just posted the bounty there. :)

**Comment by IanVS** [comment]
I'd be up for taking a crack at it if @nzakas marks as an accepted enhancement issue.

**Comment by demisx** [comment]
+1

**Comment by nzakas** [comment]
If like to see a proposal for an option name that makes sense before committing to accepting this.

**Comment by feross** [comment]
I propose we call the new option `"except-undefined"`.

The `"one-var"` rule is for requiring one `var` statement per scope. The options could be:

- `"always"` - Require one `var` statement per scope.
- `"except-undefined"` - Require one `var` statement per scope, *except when all variables are `undefined`*.
- `"never"` - Disallow multiple `var` declaration.

**Comment by IanVS** [comment]
I believe the option should be camelCase, so it would be `exceptUndefined`

A reasonable plain-English reading of the rule would be: "Require `one-var`iable per line `exceptUndefined` variables, of which there can be multiple.

**Comment by nzakas** [comment]
@IanVS I agree with your line of thinking. The word "initialized" is the key here. Maybe the option can be called "uninitialized"?

**Comment by IanVS** [comment]
@nzakas That would be simpler, but is it clear enough?  `always` and `never` both answer the question "when?" so it seems to me like any other options which are added should answer that same question.  Also, if it is only called `uninitialized`, the behavior for initialized variables would be somewhat undefined and left up to assumption.  So, I think `onlyUnitialized`, while being somewhat longwinded, is also the most clear.  

All that said, either would work fine as long as the docs are clear.  What's your preference?

**Comment by nzakas** [comment]
I don't think the word "only" adds any clarity. To use your argument, it doesn't answer "when?" Maybe "always-unitialized" is what we are after?

**Comment by IanVS** [comment]
The full name of the rule is "Require Just One var Statement Per Scope."  The option being proposed here would have a full name like "Always for uninitialized, never for initialized"  There are the two parts to it.  That is why I thought "only" is a good way to combine them.

**Comment by nzakas** [comment]
Good point. Maybe we are thinking about this wrong. Perhaps what we really need is:

```js
"one-var": [2, {
    initialized: "always",
    uninitialized: " never"
}]
```

So the first option could potentially be either a string or an object (though not sure how that would affect upcoming option validation work - ping @btmills )

**Comment by btmills** [comment]
> though not sure how that would affect upcoming option validation work

No worries from a validation perspective. `no-unused-vars` already does something similar, and I've run into no issues there.

**Comment by IanVS** [comment]
So, has the naming been decided, then?

**Comment by Flet** [comment]
/me waiting for acceptance so bounty can be added :dollar:

**Comment by nzakas** [comment]
@IanVS I believe so.

@Flet bounty away :)

**Comment by IanVS** [comment]
@nzakas I see one more wrinkle now that I look at the code.  There are currently options for `var`, `let`, and `const`, such as:

```
{ var: "always", let: "never", const: "always" }
```

So, this is getting a little messy, but should nested options be supported like so?

```
{ 
  var: {
    initialized: "never",
    uninitialized: "always"
  },
  let: {
    initialized: "always",
    uninitialized: "never"
  },
  const: "always"
}
```

I can't imagine too many people using an option like this, but that would give maximum flexibility.

**Comment by nzakas** [comment]
I'm not sure we need to go that far for this.

**Comment by IanVS** [comment]
OK good. :relieved:    

So if `initialized` and/or `uninitialized` are specified, they will apply across `var`, `let`, and `const`.

**Comment by nzakas** [comment]
Yup

**Comment by feross** [comment]
@IanVS - What is your bitcoin address? I want to send you $20, as promised. This issue wasn't listed on BountySource for some reason.

**Comment by nzakas** [comment]
BountySource is being really flaky lately, I've contacted support.

**Comment by feross** [comment]
@IanVS - ping?

**Comment by IanVS** [comment]
Thanks @feross for your kind offer.   Could I request that instead of paying me, can you add $20 amongst the `1.0.0` milestone issues?

**Comment by feross** [comment]
@IanVS Okay, generous of you :-) Added $20 to https://github.com/eslint/eslint/issues/1898

**Comment by jonathanKingston** [comment]
@IanVS thanks for this, added my previously earned $15 from ESLint to #1898
